### ISSUE TITLE
Vagrant build on Windows is broken

### ISSUE THREAD

**Issue by ghost** [freeciv/freeciv-web#168]
Currently, building Freeciv-web on Vagrant doesn't work on Windows very well. There are multiple things which needs fixing to get it running. I wonder if it is related to the Unified install script or some of the more recent commits after than.

Previously, starting Freeciv-web on Windows simply involved a git checkout, then 'vagrant up' and that was all which was required to get Freeciv-web running. Now, there is a lot of editing config files and fixing bugs based on error messages.

This is the last error message I got before giving up:
$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'ubuntu-bionic64'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: freeciv-web_default_1530961692307_45366
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: You are trying to forward to privileged ports (ports <= 1024). Most
==> default: operating systems restrict this to only privileged process (typically
==> default: processes running as an administrative user). This is a warning in case
==> default: the port forwarding doesn't work. If any problems occur, please try a
==> default: port higher than 1024.
==> default: Forwarding ports...
    default: 80 (guest) => 80 (host) (adapter 1)
    default: 443 (guest) => 443 (host) (adapter 1)
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection reset. Retrying...
    default: Warning: Connection aborted. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Connection aborted. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Connection aborted. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Connection aborted. Retrying...
    default: Warning: Connection reset. Retrying...
    default: Warning: Connection aborted. Retrying...
    default:
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default:
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Mounting shared folders...
    default: /vagrant => D:/freeciv-web
==> default: Running provisioner: systemd-resolved workaround (shell)...
    default: Running: inline script
==> default: Running provisioner: bootstrap (shell)...
    default: Running: inline script
    default: =================================
    default: Running Freeciv-web setup script.
    default: =================================
    default:
    default: Linux ubuntu-bionic 4.15.0-12-generic #13-Ubuntu SMP Thu Mar 8 06:24:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
    default: basedir: /vagrant
    default: No LSB modules are available.
    default: Distributor ID:    Ubuntu
    default: Description:       Ubuntu Bionic Beaver (development branch)
    default: Release:   18.04
    default: Codename:  bionic
: No such file or directoryts/install/install.sh: line 215: /vagrant/scripts/install/deb.sh
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

**Comment by lonemadmax** [comment]
The editing of config files is indeed due to the new scripts. Now you have to create and edit scripts/configuration.sh and freeciv-web/src/main/webapp/WEB-INF/config.properties, be it a vagrant or direct install, be it for testing(*) or for deployment. One day it will be just one config file.

And that should be all. If I understand it correctly, the error you are getting says there is no deb.sh script in the same directory as the install.sh script that is running. So either you inadvertently removed it when trying to solve other errors, or something weird is going on.

[*] Passing --mode=TEST to the install script copies the distributed files into those if they don't exist. Maybe we should make that the default for vagrant as it is for travis?

**Comment by ghost** [comment]
> Maybe we should make that the default for vagrant as it is for travis?

Yes, please.

>And that should be all. If I understand it correctly, the error you are getting says there is no deb.sh script in the same directory as the install.sh script that is running. So either you inadvertently removed it when trying to solve other errors, or something weird is going on.


The deb.sh file is there. So it seems a problem on line 215 in install.sh
https://github.com/freeciv/freeciv-web/blob/develop/scripts/install/install.sh#L215

This is what I did:
Copy and edit scripts/configuration.sh and freeciv-web/src/main/webapp/WEB-INF/config.properties
Then run 'vagrant up', and get this output:

    default: =================================
    default: Linux ubuntu-bionic 4.15.0-12-generic #13-Ubuntu SMP Thu Mar 8 06:24:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
    default: basedir: /vagrant
    default: No LSB modules are available.
    default: Distributor ID:    Ubuntu
    default: Description:       Ubuntu Bionic Beaver (development branch)
    default: Release:   18.04
    default: Codename:  bionic
    default: /vagrant/scripts/configuration.sh: line 5: $'\r': command not found
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

Then I run:
cd /d/freeciv-web/scripts
$ dos2unix.exe configuration.sh
dos2unix: converting file configuration.sh to Unix format...

Then I get this output:
....
    default: /vagrant => D:/freeciv-web
==> default: Running provisioner: systemd-resolved workaround (shell)...
    default: Running: inline script
==> default: Running provisioner: bootstrap (shell)...
    default: Running: inline script
    default: =================================
    default: Running Freeciv-web setup script.
    default: =================================
    default:
    default: Linux ubuntu-bionic 4.15.0-12-generic #13-Ubuntu SMP Thu Mar 8 06:24:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
    default: basedir: /vagrant
    default: No LSB modules are available.
    default: Distributor ID:    Ubuntu
    default: Description:       Ubuntu Bionic Beaver (development branch)
    default: Release:   18.04
    default: Codename:  bionic
: No such file or directoryts/install/install.sh: line 215: /vagrant/scripts/install/deb.sh
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

Then I do:
/vagrant/scripts/install$ vim ./install.sh

I replace this:
 "${basedir}/scripts/install/${FCW_INSTALL_SCRIPT}"

with this:

. "${basedir}/scripts/install/deb.sh"

Then the install continues correctly.

I think it would be best if you could please test the Vagrant build of Freeciv-web on a Windows 10 system.

**Comment by ghost** [comment]
Also, after the install has completed, then Freeciv-web should be running, but it is not, in particular nginx is not running.
This is the log output:


    default: Freeciv-web installed!
    default: Some components have been installed outside the package managers:
    default:   mysql-connector-python
    default:   tornado
    default:
    default: You may want to personalize some things before starting it:
    default: - Replace the scripts to start/stop dependency services in scripts.
    default: - Configure the game server in publite2/settings.ini, and update/create
    default:   pubscript_*.serv to suit your needs.
    default: - Set the mail account data for pbem games in pbem/settings.ini, and the
    default:   templates for the messages in pbem/email_template* (at least the URL).
    default: - Users for tomcat-admin web interface.
    default: - Point /etc/nginx/snippets/freeciv-web-ssl.conf to your certificate and key,
    default:   or remove SSL support from /etc/nginx/sites-available/freeciv-web.
    default:
    default: Then run scripts/start-freeciv-web.sh and enjoy!
==> default: Running provisioner: startup (shell)...
    default: Running: inline script
    default: Starting up Freeciv-web: nginx, tomcat, publite2, freeciv-proxy.
    default: Starting nginx first.
    default: Please enter root password:
    default: Job for nginx.service failed because the control process exited with error code.
    default: See "systemctl status nginx.service" and "journalctl -xe" for details.
    default: Starting up Tomcat
    default: Will sleep for 8 seconds, then do a status test...
    default: Starting Freeciv-Earth-mapgen.
    default: starting meta-stats.py
    default: Starting publite2
    default: Publite2 started
    default: Starting Freeciv-PBEM
    default: Checking that Freeciv-web is running correctly...
    default:
    default:
    default: checking nginx on http://localhost
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: nginx not running!
    default:
    default: --------------------------------
    default: checking Tomcat on http://localhost:8080/
    default: Tomcat OK!
    default: checking freeciv-web on Tomcat on http://localhost:8080/freeciv-web
    default: freeciv-web webapp OK!
    default:
    default: --------------------------------
    default: checking freeciv-proxy directly on http://localhost:7001/status
    default: freeciv-proxy OK!
    default:
    default: --------------------------------
    default: checking freeciv-proxy through nginx on http://localhost/civsocket/7001/status
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: freeciv-proxy not running correctly through nginx!
    default:
    default: --------------------------------
    default: testing WebSocket connection directly to Tornado
    default: This should show "HTTP/1.1 101 Switching Protocols" and then timeout after 2 seconds.
    default:
    default:
    default: %
    default:
    default: T
    default: o
    default: t
    default: a
    default: l
    default:
    default:
    default:
    default:
    default: %
    default:
    default: R
    default: e
    default: c
    default: e
    default: i
    default: v
    default: e
    default: d
    default:
    default: %
    default:
    default: X
    default: f
    default: e
    default: r
    default: d
    default:
    default:
    default: A
    default: v
    default: e
    default: r
    default: a
    default: g
    default: e
    default:
    default: S
    default: p
    default: e
    default: e
    default: d
    default:
    default:
    default:
    default: T
    default: i
    default: m
    default: e
    default:
    default:
    default:   Time     Time  Current
    default:                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
    default: HTTP/1.1 101 Switching Protocols
    default: Server: TornadoServer/4.5.3
    default: Date: Sat, 07 Jul 2018 20:37:07 GMT
    default: Upgrade: websocket
    default: Connection: Upgrade
    default: Sec-Websocket-Accept: KXGdZMoi2440O31GaOAuKsZfThY=
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
    default: curl: (28) Operation timed out after 2001 milliseconds with 0 bytes received
    default: testing WebSocket connection through nginx
    default: This should show "HTTP/1.1 101 Switching Protocols" and then timeout after 2 seconds.
    default:
    default:
    default: %
    default:
    default: T
    default: o
    default: t
    default: a
    default: l
    default:
    default:
    default:
    default:
    default: %
    default:  Received % Xferd  Average Speed   Time    Time     Time  Current
    default:                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
    default:
    default: --------------------------------
    default:  checking freeciv-web / publite2
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: freeciv-web OK!
    default:
    default: --------------------------------
    default: checking that webclient.min.js has been generated...
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: webclient.min.js is not OK
    default:
    default: --------------------------------
    default: checking that tileset has been generated...
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: tileset is not OK
    default:
    default: --------------------------------
    default: checking metaserver on PHP server...
    default: curl: (7) Failed to connect to localhost port 80: Connection refused
    default: metaserver is not OK
    default:
    default: --------------------------------
    default: Check of Freeciv-web is done!

**Comment by lonemadmax** [comment]
Unfortunately I can't test anything in any Windows system.

You are right I missed the nginx stuff. As pointed out in the instructions:

> - Point /etc/nginx/snippets/freeciv-web-ssl.conf to your certificate and key,
>    or remove SSL support from /etc/nginx/sites-available/freeciv-web.

Again solved in test mode, as a self-signed certificate is generated in that case.

Did you just copy the file for configuration.sh or did you also edit it? Given the message and specially your solution, it points to the difference in line endings in Windows, which I guess a Windows editor would change.

The other one has me baffled, though. The substitution you are doing by hand is the same the script should be doing, and it's what the error message is saying: can't find /vagrant/scripts/install/deb.sh. Could be another CR? Can you check the line endings in scripts/install/systems? If it's the DOS type, could you get a list of all such files? I guess I'll have to add a dos2unix call at the beginning of the script for all those files even if I don't like it.

**Comment by ghost** [comment]
Thanks for the help so far!

> You are right I missed the nginx stuff. As pointed out in the instructions:

Can you please update the instructions on "Running Freeciv-web with Vagrant on VirtualBox" here also? 
https://github.com/freeciv/freeciv-web/blob/develop/README.md 
It should contain the instructions required to run Freeciv-web on Vagrant.

> Point /etc/nginx/snippets/freeciv-web-ssl.conf to your certificate and key,

By default, I think the Vagrant build process should create a working nginx configuration which uses only HTTP (no SSL setup, and no manual steps required).

> Could be another CR? Can you check the line endings in scripts/install/systems? 

$ file systems
systems: ASCII text, with CRLF line terminators

Running dos2unix on the 'systems' file fixed this and the install continued correctly.

So in summary, can you please update so that:
- Run dos2unix is run on systems and configuration.sh
- Make nginx work by default on HTTP without any manual setup. HTTPS can be setup by the user manually later if required.
- Update so that scripts/configuration.sh and freeciv-web/src/main/webapp/WEB-INF/config.properties are copied automatically if they don't already exist, so that a simple 'vagrant up' will create a running Freeciv-web system without any manual steps.
- Update README with instructions for running on Vagrant on Windows.

**Comment by ghost** [comment]
Thanks!

**Comment by andreasrosdalw** [comment]
Unfortunately I still get this problem.  Running dos2unix on configuration.sh and systems solves the problem. Perhaps we should just run dos2unix on the files, instead of reencoding the line endings?

This is the log output I get on the latest git version of Freeciv-web:

==> default: Running provisioner: systemd-resolved workaround (shell)...
    default: Running: inline script
==> default: Running provisioner: bootstrap (shell)...
    default: Running: inline script
    default: =================================
    default: Running Freeciv-web setup script.
    default: =================================
    default:
    default: Linux ubuntu-bionic 4.15.0-12-generic #13-Ubuntu SMP Thu Mar 8 06:24:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
    default: basedir: /vagrant
    default: No LSB modules are available.
    default: Distributor ID:    Ubuntu
    default: Description:       Ubuntu Bionic Beaver (development branch)
    default: Release:   18.04
    default: Codename:  bionic
    default: Default config parameters used
    default: Default config.properties used
    default: /vagrant/scripts/configuration.sh: line 5: $'\r': command not found
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.


I don't think the changes made in this commit solved it:
https://github.com/freeciv/freeciv-web/commit/3b53280e17e73e796cf8a8530f245242606ae0d1

**Comment by lonemadmax** [comment]
You can't count on having dos2unix at that point, even if it is in the current image. And it would do the same that sed is doing, so it wouldn't help (I mean, just changing the sed incantation for a dos2unix run).

3b53280e17e73e796cf8a8530f245242606ae0d1 should have solved the problem, but it clearly hasn't, so let's entertain a second hypothesis: I'm stupid. What the change does is convert line endings only when there's a configuration.sh. If there's no such file, it just copies the distributed one, because, well, the problem can only be caused by those pesky Windows users opening configuration.sh in notepad, of course.

Or can it? Let's say git itself converts line endings for text files (for some of them, following this line it would convert configuration.sh.dist but not install.sh) and you are doing a clean install. You'd have CR in configuration.sh.dist, it would be copied as-is to configuration.sh, which would later be sourced and make the house burn. Running install a second time would get the copy, think it may be dealing with a Windows heretic, transform line endings and install correctly.

The content of .gitattributes and a quick visit to the manual also point in that direction. I may change that file when I'm more confident in its effects on existing files.

So, was it a clean install with defaults (I guess it was)? Do you have a CRLF configuration.sh.dist (I guess you have and it was automatic, not your editing)? You seem to have transformed the files yourself and finished installation. If I'm right this time, rerunning install.sh after the error would have seen the new configuration.sh and exercise the transformation path, finishing installation, even without the dos2unix assistance. If you have not touched anything after the error, could you delete configuration.sh and try again with b32cc5ecb51f4c5672e4005bc7678528181e1aaf?

**Comment by andreasrosdalw** [comment]
I used the default settings, and didn't copy or modify the configuration-sh file.
So "Default config.properties used" came in the logs as in my previous post.
However, in this default case the line ending are not modified by sed, because the else condition is not run in that case.  
Here:
https://github.com/freeciv/freeciv-web/blob/develop/scripts/install/install.sh#L176

So chaning the code to always run sed, also in the default case, will solve this for me, I think.

> So, was it a clean install with defaults (I guess it was)? 

Yes, it was a clean install with defaults.

**Comment by lonemadmax** [comment]
To think that I included that message so that the logs showed whether the defaults were used or not without having to ask the reporter of a bug... I may have to start using checklists and doing memory and attention exercises.

Thank you for your patience.

**Comment by andreasrosdalw** [comment]
Thanks!  I tested this, and needed one more change. See the pull-request.
